### ISSUE TITLE
BUG: special.nctdtr: incorrect results

### ISSUE THREAD

**Issue by Carreau** [scipy/scipy#8344]
Reported to me by @jb-leger who had weird issue with `scipy.stats.nct.logcdf`, and tracked down a bit more.

### Reproducing code example:
```
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

import scipy.special as sc 
x = np.linspace(-1,1, 100)
df=3000
nc=3
y = sc.nctdtr(df, nc, x)
plt.plot(x,y)
```

![download-3](https://user-images.githubusercontent.com/335567/35547605-cd46655a-052e-11e8-88ff-121f7e3a9881.png)

### Error message:

No errors. Plot just seem wrong with a step which should not be here. Or at least should raise an error message if parameters are outside supported range.

### Scipy/Numpy/Python version information:
```
SciPy: 1.1.0.dev0+beb258e  (Jan 24 2018)
NumPy:  1.13.3 
Python:  sys.version_info(major=3, minor=6, micro=3, releaselevel='final', serial=0)
```

I believe that end up calling `scipy/special/cdflib/cumtnc.f`

**Comment by WarrenWeckesser** [comment]
Definitely a bug.

The [wikipedia article on the noncentral T distribution](https://en.wikipedia.org/wiki/Noncentral_t-distribution) includes a formula for the CDF that begins with Φ(-μ), where Φ(x) is the normal distribution CDF and μ is the noncentrality parameter.  The second part of the formula is a series.  It looks like `nctdtr(df, nc, t)` is losing the series when `t` is small enough:

```
In [136]: from scipy.special import nctdtr, ndtr

In [137]: nctdtr(3000, 3, 0.1)
Out[137]: 0.0013498980316300946

In [138]: ndtr(-3)
Out[138]: 0.0013498980316300933 
```

The problem appears to be independent of `df`.  For example, a similar plot is generated with `df` set to 1.

**Comment by WarrenWeckesser** [comment]
Apparently the test at https://github.com/scipy/scipy/blob/327ef9c6ef9697d5ed16af14a1181390822f0824/scipy/special/cdflib/cumtnc.f#L159 is too loose.  If I comment out that `IF` statement completely, so the code always continues on to the subequenct calculation, then `nctdtr` appears to generate reasonable values near t = 0.

**Comment by WarrenWeckesser** [comment]
Commenting out that short-cut is not a viable solution, because doing so appears to result in an infinite loop when the tests in `test_cdflib.py` are run.

A correct solution will probably require digging in to the code in `cumtnc.f` a bit more.  I'm not going to be able to spend more time on it at the moment, so if someone wants to dive in, please do.

**Comment by jb-leger** [comment]
I do not know if it is why tests are going in infinite loops, but this short-cut is *not* a correct solution:

```
In [36]: sc.nctdtr(10, 3, 1e-7)
Out[36]: 0.0013499088153805516

In [37]: sc.nctdtr(10, 3, 1e-8)
Out[37]: 1.0

In [38]: sc.nctdtr(10, 3, 1e-9)
Out[38]: 1.0

In [39]: sc.nctdtr(10, 3, 0)
Out[39]: 0.0013498980316300946
```

Thanks @Carreau, I was still in investigations. I already had the short-cut idea, a bad one.

**Comment by josef-pkt** [comment]
I'm not trying to understand the code.

However, it seems to me the normal approximation around 0 should still have t in it, which would replace the flat part by the local normal distribution behavior.

If df is large, then the neighborhood where nct is close to normal is also larger.
If df is small, e.g. df <=1 , then the neighborhood with good approximation is smaller than the currently flat part (tiny too large ?)

Maybe something like this

```
      IF ((dum1+dum2).LT.tiny) THEN
          CALL cumnor(t-pnonc,cum,ccum)
          RETURN
```

```
import scipy.special as sc 
from scipy import stats

x = np.linspace(-1,1, 100)
df=3   
nc=3
y = sc.nctdtr(df, nc, x)
yn = sc.ndtr(x-nc)
plt.plot(x,y)
plt.plot(x,yn)
plt.show()
```

**Comment by josef-pkt** [comment]
It's possible to stretch the normal approximation a bit for small df to get closer at the kinks, but it doesn't fully work with only one scaling factor.

```
x = np.linspace(-0.2,0.2, 100)
df=1
nc=3
y = sc.nctdtr(df, nc, x)
yn = sc.ndtr((x*0.8-nc))
plt.plot(x,y)
plt.plot(x,yn)
plt.show()
```

**Comment by jb-leger** [comment]
So, here my investigations.

# Initial problem.

```python
import numpy as np
import matplotlib.pyplot as plt

import scipy.special as sc
x = np.linspace(-1,1, 100)
df=10
nc=3
y = sc.nctdtr(df, nc, x)
plt.plot(x,y)
```

![sp_nct](https://user-images.githubusercontent.com/24868098/36047612-e9586868-0ddc-11e8-9a0b-0fb399e1da33.png)

As @WarrenWeckesser mentioned before, the problem was the check in the file `scipy/special/cdflib/cumtnc.f`:

```fortran
     IF ((dum1+dum2).LT.tiny) THEN
         CALL cumnor(-pnonc,cum,ccum)
         RETURN

     END IF
```

So I tried [this patch](https://github.com/jb-leger/scipy/commit/a7e057fb58e1772c81f5df7f2f9c8fa169486e83) (please note I do not want to sumit this patch, I did'nt read `CONTRIBUTING`, it is only exploratory work.

Ok, this works file:

![sp_nct0](https://user-images.githubusercontent.com/24868098/36048109-6abf2094-0dde-11e8-8ae2-229cf1b09a13.png)

# Ok, but it is not a so good idea

Firstly, one problem happen:

```python
import numpy as np
import matplotlib.pyplot as plt

import scipy.special as sc
x1 = np.exp(np.linspace(-23,0,100))
x = np.concatenate((-x1[::-1],(0,),x1))
df=10
nc=1
y = sc.nctdtr(df, nc, x)
plt.plot(x,y)
```

We have:
![sp_nct1](https://user-images.githubusercontent.com/24868098/36048190-a931f874-0dde-11e8-9ee8-236197cf5793.png)

Let use the zoom:
![sp_nct2](https://user-images.githubusercontent.com/24868098/36048204-ba3e4726-0dde-11e8-9a4d-d731c09603d8.png)

In fact, the code fail for very small values of `t`. This time, in `scipy/special/cdflib/cumtnc.f` we have:

```fortran
     x = df/ (df+t2)
     omx = one - x
```

If `t2` is small (`t2` is `t` square) w.r.t. `df`, then `x` should be near `1` (but not `1`) and `omx` near `0`, but with _fp_ computation, `x` is rounded to exactly `1` and `omx` exactly `0`. So when `x>½` I propose to compute precisely `omx` and deduce `x` with `omx`.

```fortran
     IF (df.LE.t2) THEN
         x = df/ (df+t2)
         omx = one - x
     ELSE
         omx = t2/ (df+t2)
         x = one - omx
     END IF
```

It is what I did with [this patch](https://github.com/jb-leger/scipy/commit/f381bafa6ae7c4c3160df9f4c32a8459be967093) (again, it is exploratory work).

So we have:

![sp_nct3](https://user-images.githubusercontent.com/24868098/36048494-b9f66d92-0ddf-11e8-94e7-cca66996497a.png)

and

![sp_nct4](https://user-images.githubusercontent.com/24868098/36048506-c3723fb8-0ddf-11e8-9060-60f8fb1201f5.png)

Ok, this problem, solved.

# Another one

@WarrenWeckesser had a _infinite_ loop in tests. Ok, I had a look. This happens with the function `sc.nctdtrinc` (which compute the `nc` given the other parameters, by inverting the cdf function):

```python
import scipy.special as sc
df,nc,t = 10,3,1
pp = sc.nctdtr(df, nc, t)
sc.nctdtrinc(df, pp, t) # I have to send a SIGKILL to python to stop that
```

Ok, more precisely, we have in `scipy/special/cdflib/cdftnc.f`:

```fortran
      ELSE IF ((4).EQ. (which)) THEN
          pnonc = 5.0D0
          CALL dstinv(-tent6,tent6,0.5D0,0.5D0,5.0D0,atol,tol)
          status = 0
          CALL dinvr(status,pnonc,fx,qleft,qhi)
  210     IF (.NOT. (status.EQ.1)) GO TO 220
          CALL cumtnc(t,df,pnonc,cum,ccum)
          fx = cum - p
          CALL dinvr(status,pnonc,fx,qleft,qhi)
          GO TO 210
```

`dinvr` set very negative value to `pnonc` (the `nc` variable in python), and call the `cumtnc`. In fact, problems do not happens to very negative values:

```python
In [31]: import scipy.special as sc
    ...: df,t = 10,1

In [32]: %time pp = sc.nctdtr(df, -1000, t)
CPU times: user 61.2 ms, sys: 0 ns, total: 61.2 ms
Wall time: 61 ms

In [33]: %time pp = sc.nctdtr(df, -2000, t)
CPU times: user 232 ms, sys: 0 ns, total: 232 ms
Wall time: 231 ms

In [34]: %time pp = sc.nctdtr(df, -10000, t)
CPU times: user 5.96 s, sys: 10.8 ms, total: 5.97 s
Wall time: 5.96 s
```

Ok, for very negative `nc` (when `t` positive), we can maybe assume something, but if we look carefully, we have strange behavior even when `nc` is not so small:

```python
import numpy as np
import matplotlib.pyplot as plt

import scipy.special as sc
df,t = 45,10
xnc = np.linspace(-5,5,1001)
y = 1-sc.nctdtr(df,xnc,t)
plt.plot(xnc,np.log(1e-15+y)-np.log(1e-15))
```

We have:
![sp_nct9](https://user-images.githubusercontent.com/24868098/36050106-73345148-0de5-11e8-953c-e2687e472752.png)

We see the `0` value which is not computed with the serie but with the traditional student distrib.

I do not exactly understand the behavior of the used serie (which is not the same as wikipedia article).

**Comment by dschmitz89** [comment]
This issue is fixed for `stats`: switching `nctdtr` with `stats.nct.cdf` results in

![image](https://github.com/user-attachments/assets/dee5cb44-9e8c-4d85-91fd-6f440a85f640)

<details><summary>Details</summary>
<p>

```python
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
x = np.linspace(-1,1, 100)
df=3000
nc=3
y = stats.nct.cdf(x, df, nc)
plt.plot(x,y)
``` 

</p>
</details>

**Comment by Carreau** [comment]
Thanks, that is useful, I can say that this has been fixed between 1.9.0 and 1.10.0.

I don't know the difference between `special` and `stats`, but the special issue is still here.

I'm going to guess this was fixed by #16591 ; should special be an alias for the stats one and/or be deprecated ? It does return wrong values.

**Comment by lucascolley** [comment]
> I'm going to guess this was fixed by #16591 ; should special be an alias for the stats one and/or be deprecated ? It does return wrong values.

Since `special` is working towards CUDA support, I think there is hesitation to depend on Boost, although I don't think it has been completely ruled out. cc @steppi

**Comment by Carreau** [comment]
I think it might be worth it ton consider deprecating and removing (potentially reading later), with a suggestion to use stats if the special computation is wrong.

**Comment by steppi** [comment]
> Since `special` is working towards CUDA support, I think there is hesitation to depend on Boost, although I don't think it has been completely ruled out for the future. cc @steppi

It appears the Boost devs are open to adding CUDA support, see https://github.com/scipy/scipy/pull/20785#issuecomment-2223387794, and if so, I don't think there needs to be any hesitation about using Boost.

**Comment by dschmitz89** [comment]
This was fixed in #21505.

**Comment by dschmitz89** [comment]
This could be fixed in the same way as #21505.

**Comment by mdhaber** [comment]
Yup, that's the plan!
